packages/core/src/utils/storage/google-storage.ts

```diff
@@ -0,0 +1,33 @@
+import { Storage } from '@google-cloud/storage';
+
+import type { UploadFile } from './types.js';
+
+const defaultPublicDomain = 'storage.googleapis.com';
+
+export const buildGoogleStorage = (projectId: string, keyFilename: string, bucketName: string) => {
+  const storage = new Storage({ projectId, keyFilename });
+  const bucket = storage.bucket(bucketName);
+
+  const uploadFile: UploadFile = async (
+    data: Buffer,
+    objectKey: string,
+    { contentType, publicUrl } = {}
+  ) => {
+    const file = bucket.file(objectKey);
+
+    await file.save(data, {
+      contentType,
+      public: Boolean(publicUrl),
+    });
+
+    if (publicUrl) {
+      return { url: `${publicUrl}/${objectKey}` };
+    }
+
+    return {
+      url: `https://${defaultPublicDomain}/${bucketName}/${objectKey}`,
+    };
```

Comment:
Same as above, we can handle the URL concatenation safely with `appendPath`. E.g.

```ts
const defaultPublicDomain = 'https://storage.googleapis.com';
...
return {
  url:  appendPath(defaultPublicDomain, bucketName, objectKey),
}
```